core/app/models/spree/discounts/order_updater.rb

```diff
@@ -10,7 +10,57 @@ def initialize(order)
       end
 
       def call
-        order
+        update_line_items
+        update_shipments
+      end
+
+      private
+
+      def update_line_items
+        line_item_promotions = promotions.select do |promotion|
+          promotion.actions.any? { |promotion_action| promotion_action.discountable_class == Spree::LineItem }
+        end
+        line_item_promo_updater = LineItemUpdater.new(promotions: line_item_promotions)
+        order.line_items.each { |line_item| line_item_promo_updater.call(line_item) }
+      end
+
+      def update_shipments
+        shipment_promotions = promotions.select do |promotion|
+          promotion.actions.any? { |promotion_action| promotion_action.discountable_class == Spree::Shipment }
+        end
+        shipment_promo_updater = ShipmentUpdater.new(promotions: shipment_promotions)
+        order.shipments.each { |shipment| shipment_promo_updater.call(shipment) }
+      end
+
+      def promotions
+        @_promotions ||= begin
+          preloader = ActiveRecord::Associations::Preloader.new
+          (connected_order_promotions | sale_promotions).select do |promotion|
+            promotion.activatable?(order)
+          end.map do |promotion|
+            preloader.preload(promotion.rules.select { |r| r.type == "Spree::Promotion::Rules::Product" }, :products)
+            preloader.preload(promotion.rules.select { |r| r.type == "Spree::Promotion::Rules::Store" }, :stores)
+            preloader.preload(promotion.rules.select { |r| r.type == "Spree::Promotion::Rules::Taxon" }, :taxons)
+            preloader.preload(promotion.rules.select { |r| r.type == "Spree::Promotion::Rules::User" }, :users)
```

Comment:
Thought for later: Can we make this configurable for stores having own rules with related objects? an `AlchemyPage` rule for example ;)